File: Go/test_harness.go
package test_harness

import ( "fmt"; "strconv" )

type TestFunc func()

var nPass int32
var nFail int32
var nTotalPass int32
var nTotalFail int32

func summary() {
  fmt.Printf("   Passed: %v\n   Failed: %v\n\n",nPass,nFail)
}

func FinalSummary() {
  if nFail > 0 || nPass > 0 {
    summary();
  }
  nTotalPass+=nPass
  nTotalFail+=nFail
  fmt.Printf("Total Passed: %v\nTotal Failed: %v\n",nTotalPass,nTotalFail)
}

func TestCase(msg string) {
  if nFail > 0 || nPass > 0 {
    summary();
  }
  nTotalPass+=nPass
  nTotalFail+=nFail
  nPass=0
  nFail=0
  fmt.Println(msg)
}

func Test(msg string,rc bool) {
  var result string
  if rc {
    result="pass"
    nPass++
  } else {
    result="fail"
    nFail++
  }
  fmt.Printf("   %v ... %v\n",msg,result)
}

//var Tests []TestFunc

func RunTests(argv [] string,tests [] TestFunc) {
  if len(argv)> 0 {
    nTests := len(tests)
    for _,istr := range argv {
      i,_ := strconv.Atoi(istr)
      if i<nTests  {
        tests[i]()
      } else {
        fmt.Println("No test for ",i)
      }
    }
  } else {
    for _,testfunc := range tests {
      testfunc()
    }
  }
}

File: JavaScript/test_harness.js
class TestHarness {

  constructor() {
    this.nPass=0;
    this.nFail=0;
    this.nTotalPass=0;
    this.nTotalFail=0;
  }

  summary()
  {
    console.log("  Passed: ",this.nPass,"\n  Failed: ",this.nFail,"\n");
  }

  final_summary()
  {
    if(this.nPass != 0 || this.nFail != 0) {
      this.summary();
    }
    this.nTotalPass += this.nPass;
    this.nTotalFail += this.nFail;
    console.log("  Total Passed: ",this.nTotalPass,"\n  Failed: ",this.nTotalFail);
  }

  test_case(str)
  {
    if(this.nPass != 0 || this.nFail != 0) {
      this.summary();
    }
    console.log("\n",str);
    this.nTotalPass += this.nPass;
    this.nTotalFail += this.nFail;
    this.nPass=0;
    this.nFail=0;
  }

  test(str,result)
  {
    var result_str;
    if(result) {
      this.nPass++;
      result_str="Pass"
    } else {
      this.nFail++;
      result_str="Fail"
    }
    console.log("  ",str," ... ",result_str);
  }
}

module.exports = TestHarness;
